u2f-ref-code/java/tests/com/google/u2f/server/impl/U2FServerReferenceImplTest.java

```diff
@@ -236,6 +237,123 @@ public void testProcessSignResponse_badOrigin() throws U2FException {
+                         BROWSER_DATA_TRANSFER_ACCESS_BASE64, 
+                         SESSION_ID);
+
+    u2fServer.processSignResponse(transferAccessResponse, ENROLLMENT_TIME);
+    
+    verify(mockDataStore).removeSecurityKey(eq(ACCOUNT_NAME),
+        AdditionalMatchers.aryEq(TRANSFER_ACCESS_PUBLIC_KEY_A_HEX));
+    verify(mockDataStore).addSecurityKeyData(eq(ACCOUNT_NAME),
+        eq(new SecurityKeyData(ENROLLMENT_TIME, 
+                               null, 
+                               KEY_HANDLE_B,
+                               TRANSFER_ACCESS_PUBLIC_KEY_B_HEX, 
+                               VENDOR_CERTIFICATE, 
+                               INITIAL_COUNTER)));
+  }
+
+  @Test
+  public void testProcessSignResponse_withTransferAccess_multipleTransfers()
+      throws U2FException {
+    when(mockDataStore.getSignSessionData(SESSION_ID)).thenReturn(new SignSessionData(ACCOUNT_NAME,
+        APP_ID_SIGN, SERVER_CHALLENGE_SIGN, TRANSFER_ACCESS_PUBLIC_KEY_A_HEX));
+    u2fServer =
+        new U2FServerReferenceImpl(mockChallengeGenerator, mockDataStore, crypto, TRUSTED_DOMAINS);
+    SignResponse transferAccessResponse =
+        new SignResponse(KEY_HANDLE_A_BASE64, 
+                         TRANSFER_ACCESS_RESPONSE_A_TO_B_TO_C_TO_D_NO_USER_PRESENCE_BASE64,
+                         BROWSER_DATA_TRANSFER_ACCESS_BASE64, 
+                         SESSION_ID);
+
+    u2fServer.processSignResponse(transferAccessResponse, ENROLLMENT_TIME);
+    
+    verify(mockDataStore).removeSecurityKey(eq(ACCOUNT_NAME),
+        AdditionalMatchers.aryEq(TRANSFER_ACCESS_PUBLIC_KEY_A_HEX));
+    verify(mockDataStore).addSecurityKeyData(eq(ACCOUNT_NAME),
+        eq(new SecurityKeyData(ENROLLMENT_TIME, 
+                               null, 
+                               KEY_HANDLE_D,
+                               TRANSFER_ACCESS_PUBLIC_KEY_D_HEX, 
+                               VENDOR_CERTIFICATE, 
+                               INITIAL_COUNTER)));
+  }
+      
+  @Test
+  public void testProcessSignResponse_withTransferAccess_badOrigin() throws U2FException {
+    when(mockDataStore.getSignSessionData(SESSION_ID)).thenReturn(new SignSessionData(ACCOUNT_NAME,
+        APP_ID_SIGN, SERVER_CHALLENGE_SIGN, TRANSFER_ACCESS_PUBLIC_KEY_A_HEX));
+    u2fServer = new U2FServerReferenceImpl(mockChallengeGenerator,
+        mockDataStore, crypto, ImmutableSet.of("some-other-domain.com"));
+    SignResponse transferAccessResponse = 
+        new SignResponse(KEY_HANDLE_A_BASE64, 
+                         TRANSFER_ACCESS_RESPONSE_A_TO_B_BASE64, 
+                         BROWSER_DATA_TRANSFER_ACCESS_BASE64, 
+                         SESSION_ID);
+
+    try {
+      u2fServer.processSignResponse(transferAccessResponse, ENROLLMENT_TIME);
+      fail("expected exception, but didn't get it");
+    } catch(U2FException e) {
+      assertTrue(e.getMessage().contains("is not a recognized home origin"));
+    }
+  }  
+
```

Comment:
I didn't explicitly test for this because a wrong sequence number would be caught by a signature that doesn't check out.

For example, say the sequence numbers were (somehow) out of order, but the signatures still all checked out. Is this a problem? There is still a clear transition of trust in the attestation and authentication keys in the correct order, they've just somehow signed them with sequence numbers that aren't in order.

This would not work in our parser, but for the purposes of checking here, I don't think it is important.